File: tests/fixtures/sample_dead_guard.py
"""Fixture for testing dead-guard detection on CALLS edges.

Contains calls under ``if False:``, ``if 0:``, ``if TYPE_CHECKING:``,
and live calls as controls.  The parser should:

* **omit** CALLS edges inside the consequence (true branch) of a dead
  guard -- those edges are never emitted at all.
* **keep** CALLS edges inside the else/elif branch of a dead guard --
  those branches execute when the condition is false.
* treat ``TYPE_CHECKING`` as the typing sentinel only when it is
  imported from ``typing``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from os.path import join as pjoin  # noqa: F401


def live_helper():
    pass


def caller():
    live_helper()                    # live -- no guard

    if False:
        dead_false_call()            # noqa: F821 -- dead consequence

    if 0:
        dead_zero_call()             # noqa: F821 -- dead consequence

    if TYPE_CHECKING:
        dead_tc_call()               # noqa: F821 -- dead consequence


def else_branch():
    """Calls in the else branch of ``if False:`` are LIVE."""
    if False:
        dead_in_if()                 # noqa: F821 -- dead consequence
    else:
        live_in_else()               # live -- else branch of if False


def elif_chain():
    """Calls in elif branches of ``if False:`` are LIVE."""
    if False:
        dead_elif_consequence()      # noqa: F821 -- dead
    elif some_condition():
        live_elif_call()             # live -- elif branch
    else:
        live_final_else()            # live -- else branch


def some_condition() -> bool:
    return True

File: tests/fixtures/sample_dead_guard.c
/* Fixture for testing C/C++ dead-guard detection on CALLS edges.
 *
 * #if 0 / #elif 0 blocks are dead code -- calls inside them should be
 * omitted, even when a function definition sits inside the block.
 * #else and #elif branches of #if 0 are live -- their calls are kept.
 */

extern void live_helper(void);
extern void dead_in_if0(void);
extern void live_in_else(void);
extern void dead_in_elifblock(void);
extern void live_in_elif(void);
extern void dead_in_wrapped(void);
extern void live_in_if1(void);
extern void dead_in_elif0(void);

/* #if 0 wrapping a whole function: the preprocessor removes the
 * function entirely, so the call inside it is dead too. */
#if 0
void dead_wrapped_func(void) {
    dead_in_wrapped();   /* dead -- function is inside #if 0 */
}
#endif

void caller(void) {
    live_helper();       /* live -- no guard */

#if 0
    dead_in_if0();       /* dead -- inside #if 0 */
#else
    live_in_else();      /* live -- #else of #if 0 */
#endif

#if 0
    dead_in_elifblock(); /* dead -- inside #if 0 (elif form) */
#elif 1
    live_in_elif();      /* live -- #elif of #if 0 (regression guard) */
#endif

#if 1
    live_in_if1();       /* live -- #if 1 is taken */
#elif 0
    dead_in_elif0();     /* dead -- inside #elif 0 */
#endif
}
